### PR TITLE
Add design mode component

### PR DESCRIPTION
**Pull request by DilwoarH**
# What
Add the ability for users to edit the content of the page without using the inspector.

# Why
This will allow non-technical people to mock up page content without having knowledge/training on the inspector.

# Screenshots

### CODE REVIEW

**Comment by DilwoarH** [comment]
@alex-ju done

File: src/design-mode-component.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+'use strict';
+
+function DesignModeComponent() {
+  this.state = false;
+  chrome.runtime.onMessage.addListener(function (request) {
+    if (request.trigger == 'toggleDesignMode') {
```
Comment (ChrisBAshton) [inline src/design-mode-component.js]: Everywhere else in the file is `"`, can we make this double quotes too? (And the `"use strict"`)

File: src/design-mode-component.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      this.toggleDesignMode();
+    }
+  }.bind(this));
+}
+
+DesignModeComponent.prototype.toggleDesignMode = function () {
+  this.state = !this.state;
```
Comment (ChrisBAshton) [inline src/design-mode-component.js]: Could this be renamed to, say, `this.designModeActive`?

File: spec/javascripts/design_mode_spec.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  it("removes the banner when toggled off", function () {
+    designModeComponent.toggleDesignMode();
+    expect($bannerEl.parent()).toHaveLength(0);
+  });
+
+  it("design mode is on when toggled on", function () {
+    expect(window.document.designMode).toEqual("on");
+  });
+
+  it("design mode is on when toggled off", function () {
```
Comment (ChrisBAshton) [inline spec/javascripts/design_mode_spec.js]:
```suggestion
  it("design mode is off when toggled off", function () {
```

File: src/popup/popup.js [hunk 4]
```diff
@@ -9,6 +9,10 @@ var Popup = Popup || {};
       code: "window.highlightComponent = window.highlightComponent || new HighlightComponent; undefined;"
     });
 
+    chrome.tabs.executeScript(null, {
+      code: "window.designModeComponent = window.designModeComponent || new DesignModeComponent; undefined;"
```
Comment (ChrisBAshton) [inline src/popup/popup.js]: I know this is copied from the `highlightComponent` block, but why do we have `undefined;` at the end? Just curious!

File: src/popup/popup.js [hunk 5]
```diff
@@ -52,6 +56,16 @@ var Popup = Popup || {};
     }
   });
 
+  chrome.runtime.onMessage.addListener(function (request, _sender) {
+    if (request.action == "designModeState") {
+      var toggleLink = $('#toggle-design-mode');
+      if (request.designModeState)
+        toggleLink.text('Turn off design mode');
+      else
+        toggleLink.text('Turn on design mode');
```
Comment (ChrisBAshton) [inline src/popup/popup.js]: Again, a mix of single and double quotes, can you make all double?

File: src/popup/popup.js [hunk 6]
```diff
@@ -137,6 +155,15 @@ var Popup = Popup || {};
...
     });
+
+    $('#toggle-design-mode').on('click', function(e) {
+      e.preventDefault();
+
+      chrome.tabs.query({ active: true, currentWindow: true }, function(tabs) {
+        var govukTab = tabs[0];
+        chrome.tabs.sendMessage(govukTab.id, { trigger: 'toggleDesignMode' });
+      });
+    });
```
Comment (ChrisBAshton) [inline src/popup/popup.js]: Can probably refactor out the duplication here, as we now have 3 functions doing exactly the same thing, just different selectors and triggers.

File: src/design-mode-component.js (hunk 2 above)
Comment (DilwoarH) [reply src/design-mode-component.js]: I think state is fine as it's consistent with the other components

File: src/popup/popup.js (hunk 4 above)
Comment (DilwoarH) [reply src/popup/popup.js]: I really don't know.. there must be a reason.

Comment (alex-ju) [reply src/popup/popup.js]: I think the answer is in the comment of https://github.com/alphagov/govuk-browser-extension/commit/f445fe662925b2d048c3b2cee597d3b09ee5c8a4

Comment (ChrisBAshton) [reply src/popup/popup.js]: Ah! "* Add `undefined;` at the end of all content scripts, since they need to return something"

**Pull request merged**